lib/std/http.zig

```diff
@@ -1022,10 +1028,23 @@ pub const BodyWriter = struct {
                 },
                 else => {
                     const chunk_limit: std.Io.Limit = .limited(chunk_len - 2);
-                    const n = if (chunk_limit.subtract(w.buffered().len)) |sendfile_limit|
-                        try out.sendFileHeader(w.buffered(), file_reader, sendfile_limit.min(limit))
-                    else
-                        try out.write(chunk_limit.slice(w.buffered()));
+                    const n = blk: {
+                        if (chunk_limit.subtract(w.buffered().len)) |sendfile_limit| {
+                            break :blk out.sendFileHeader(
+                                w.buffered(),
+                                file_reader,
+                                sendfile_limit.min(limit),
+                            ) catch |err| switch (err) {
+                                error.EndOfStream => {
+                                    chunked.* = .{ .chunk_len = 0 };
```

Comment:
shouldn't this be based on observing how many bytes were read from file_reader? something like `const before = file_reader.logicalPos(); ...; const after = file_reader.logicalPos(); const n = after - before;`

and then we still would need to do the `w.consume(n)` logic

Reply:
Hmm, I guess that would potentially give us more robust assertions but I'm not sure there would be any behavioral difference. My understanding of the sendFile interface is that EndOfStream is only returned by the writer when *all* bytes have been read from the reader.

Perhaps I've misunderstood that though? I'll defer to your judgement here, I'm still learning the new interfaces (which is why I thought it would be nice to fix some bugs to enhance my understanding).

Reply:
Hmm, the doc comment for the `sendFile` function in the vtable seems to suggest that `error.EndOfStream` was not the intended way to signal stream end originally:
```
    /// Number of bytes returned may be zero, which does not indicate stream
    /// end. A subsequent call may return nonzero, or signal end of stream via
    /// `error.WriteFailed`. Caller may check `file_reader` state
    /// (`File.Reader.atEnd`) to disambiguate between a zero-length read or
    /// write, and whether the file reached the end.
```

I'm going to experiment with deleting `EndOfStream` from `Writer.FileError`.